### PR TITLE
barycentric form fixes and cleanups, with a cross-client test

### PR DESCRIPTION
**Pull request by agnxsh**
A maintenance PR to write a few portions of barycentric precompute optimization strategy differently, along with a few cleanups and then cross testing it with a Lagrange Interpolated value from Go-IPA.

More elaborately
- Avoiding typecasting and directly using `fromInt` instead converting too many times using uint64 and BigInts, this was also an issue for the CI check earlier.
- Fixing an index issue in PrecomputedWeights, the length of invertedDomain should be 2 less than that of barycentric weights, because there will be 2 such instances of 1/k and -1/k where k will essentially be 0, so we need to avoid those indexes in order to compute correctly.
- Fixing some Endianness issues, as the Verkle Spec follows only `littleEndian` when it comes to ser/de of Scalars
- Just some cosmetic things that came up now and then.

### CODE REVIEW

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 1]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
     var k {.noInit.}: Fr[Banderwagon]
-    var i_bg {.noInit.} : matchingOrderBigInt(Banderwagon)
-    i_bg.setUint(uint64(i))
-    k.fromBig(i_bg)
+    k.fromInt(i)
+    var k_inv {.noInit.} : Fr[Banderwagon]
+    k_inv.inv(k)
 
-    k.inv(k)
+    discard k
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:55]:
```suggestion
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 2]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
+    temp.diff(domain_element_Fr,i_Fr)
+    res *= temp
 
 
 func computeBarycentricCoefficients*(res_inv: var openArray[Fr[Banderwagon]], precomp: PrecomputedWeights, point : Fr[Banderwagon]) =
   ## computeBarycentricCoefficients computes the coefficients for a point `z` such that
   ## when we have a polynomial `p` in Lagrange basis, the inner product of `p` and barycentric coefficients is 
   ## equal to p(z). Here `z` is a point outside of the domain. We can also term this as Lagrange Coefficients L_i.
-  var res {.noInit.}: array[VerkleDomain, Fr[Banderwagon]]
+  var res {.noInit.} : array[VerkleDomain, Fr[Banderwagon]]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:91]:
```suggestion
  var res {.noInit.}: array[VerkleDomain, Fr[Banderwagon]]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 3]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
   for i in 0 ..< VerkleDomain:
-    var weight: Fr[Banderwagon]
+    var weight {.noInit.} : Fr[Banderwagon]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:93]:
```suggestion
    var weight {.noInit.}: Fr[Banderwagon]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 4]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
+    var i_fr {.noInit.} : Fr[Banderwagon]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:94]:
```suggestion
    var i_fr {.noInit.}: Fr[Banderwagon]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 5]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
-    var i_fr: Fr[Banderwagon]
-    i_fr.fromBig(i_bg)
+    i_fr.fromInt(i)
 
     res[i].diff(point, i_fr)
-    res[i].prod(res[i], weight)
+    res[i] *= weight
 
-  var totalProd: Fr[Banderwagon]
+  var totalProd {.noInit.} : Fr[Banderwagon]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:101]:
```suggestion
  var totalProd {.noInit.}: Fr[Banderwagon]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 6]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
   totalProd.setOne()
 
   for i in 0 ..< VerkleDomain:
-    var i_bg: matchingOrderBigInt(Banderwagon)
-    i_bg.setUint(uint64(i))
-    var i_fr: Fr[Banderwagon]
-    i_fr.fromBig(i_bg)
+    var i_fr {.noInit.} : Fr[Banderwagon]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:105]:
```suggestion
    var i_fr {.noInit.}: Fr[Banderwagon]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 7]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
+    i_fr.fromInt(i)
 
-    var tmp: Fr[Banderwagon]
+    var tmp {.noInit.} : Fr[Banderwagon]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:108]:
```suggestion
    var tmp {.noInit.}: Fr[Banderwagon]
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 8]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
+    res_inv[i] *= totalProd
 
 
 func getInvertedElement*(res: var Fr[Banderwagon], precomp : PrecomputedWeights, element : int, is_negative: bool) =
   var index: int
   index = element - 1 
 
   if is_negative:
-    var midpoint = int((len(precomp.invertedDomain) / 2)) - 1
+    var midpoint = int((len(precomp.invertedDomain) / 2))
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:124]:
```suggestion
    var midpoint = precomp.invertedDomain.len div 2
```

There is no reason to convert to float and back to int there.

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 9]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
     index = index + midpoint
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:125]: Is -1/0 properly handled? This needs a comment mentioning this gotcha and ideally a test.

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 10]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
 
   res = precomp.invertedDomain[index]
 
-
 func getWeightRatios*(result: var Fr[Banderwagon], precomp: PrecomputedWeights, numerator: var int, denominator: var int) =
   var a = precomp.barycentricWeights[numerator]
-  var midpoint = int((len(precomp.barycentricWeights) / 2)) - 1
+  var midpoint = int((len(precomp.barycentricWeights) / 2))
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:131]:
```suggestion
  var midpoint = precomp.barycentricWeights.len div 2
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 11]
```diff
@@ -38,112 +39,102 @@ func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights) =
...
   var b = precomp.barycentricWeights[denominator + midpoint]
   result.prod(a, b)
 
 
 func getBarycentricInverseWeight*(res: var Fr[Banderwagon], precomp: PrecomputedWeights, i: int) =
-  var midpoint = int((len(precomp.barycentricWeights) / 2)) - 1
+  var midpoint = int((len(precomp.barycentricWeights) / 2))
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:137]:
```suggestion
  var midpoint = precomp.barycentricWeights div 2
```

File: constantine/eth_verkle_ipa/barycentric_form.nim [hunk 12]
```diff
@@ -160,24 +151,24 @@ func absIntChecker*[int] (res: var int, x : int) =
...
+  var is_negative = true
   var y = f[index]
 
   for i in 0 ..< VerkleDomain:
-    if not(i == index).bool() == true:    
-      var denominator = i - int(index)
-      var absDenominator {.noInit.}: int
+    if i != index:    
+      var denominator = i - index
+      var absDenominator {.noInit.} : int
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/barycentric_form.nim:160]:
```suggestion
      var absDenominator {.noInit.}: int
```

File: constantine/eth_verkle_ipa/eth_verkle_constants.nim [hunk 13]
```diff
@@ -45,7 +45,7 @@ type VerkleMultiproofSerialized* = array[576, byte]
 
 type 
   PrecomputedWeights* = object
-    barycentricWeights*: array[510,Fr[Banderwagon]]
+    barycentricWeights*: array[512,Fr[Banderwagon]]
```
Comment (mratsim) [inline constantine/eth_verkle_ipa/eth_verkle_constants.nim:48]: I thought we were skipping 1/k and -1/k with k = 0?

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 14]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
             point.y = lagrange_values[k]
 
-            points[k]=point
-
-        var poly_coeff : array[VerkleDomain, Fr[Banderwagon]]
-        poly_coeff.interpolate(points, VerkleDomain)
-
-        var expected2: Fr[Banderwagon]
-        expected2.ipaEvaluate(poly_coeff, p_outside_dom, VerkleDomain)
+            discard x_fr
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:130]:
```suggestion
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 15]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
 
   test "Testing Barycentric Precompute Coefficients":
     proc testBarycentricPrecomputeCoefficients()=
-
         var p_outside_dom : Fr[Banderwagon]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:101]:
```suggestion
        var p_outside_dom: Fr[Banderwagon]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 16]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
-        var i_bg : matchingOrderBigInt(Banderwagon)
-        i_bg.setUint(uint64(3400))
-        
-        p_outside_dom.fromBig(i_bg)
-
-        var testVals: array[10,uint64] = [1,2,3,4,5,6,7,8,9,10] 
+        var testVals: array[10, int] = [1,2,3,4,5,6,7,8,9,10] 
         
-        var lagrange_values : array[256,Fr[Banderwagon]]
+        var lagrange_values : array[256, Fr[Banderwagon]]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:106]:
```suggestion
        var lagrange_values: array[256, Fr[Banderwagon]]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 17]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
         lagrange_values.testPoly256(testVals)
 
         var precomp {.noInit.}: PrecomputedWeights
-
         precomp.newPrecomputedWeights()
         
-        var bar_coeffs: array[256, Fr[Banderwagon]]
-
+        var bar_coeffs {.noInit.} : array[256, Fr[Banderwagon]]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:112]:
```suggestion
        var bar_coeffs {.noInit.}: array[256, Fr[Banderwagon]]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 18]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
         bar_coeffs.computeBarycentricCoefficients(precomp, p_outside_dom)
 
-        var expected0: Fr[Banderwagon]
-
+        var expected0 : Fr[Banderwagon]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:115]:
```suggestion
        var expected0: Fr[Banderwagon]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 19]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
         expected0.computeInnerProducts(lagrange_values, bar_coeffs)
 
-        var points: array[256, Coord]
+        var expected1 : Fr[Banderwagon]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:118]:
```suggestion
        var expected1: Fr[Banderwagon]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 20]
```diff
@@ -98,53 +98,43 @@ suite "Barycentric Form Tests":
...
+        expected1.evalOutsideDomain(precomp, lagrange_values, p_outside_dom)
+
+        var points : array[VerkleDomain, Coord]
         for k in 0 ..< 256:
-            var x : matchingOrderBigInt(Banderwagon)
-            x.setUint(uint64(k))
             var x_fr : Fr[Banderwagon]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:123]:
```suggestion
            var x_fr: Fr[Banderwagon]
```

File: tests/t_ethereum_verkle_ipa_primitives.nim [hunk 21]
```diff
@@ -452,12 +440,10 @@ suite "IPA proof tests":
...
 
         # from a shared view
-        var i_bg : matchingOrderBigInt(Banderwagon)
-        i_bg.setUint(uint64(123456789))
-        point.fromBig(i_bg)
+        point.fromInt(123456789)
 
         # from the prover's side
-        var testVals : array[9, uint64] = [1,2,3,4,5,6,7,8,9]
+        var testVals : array[9, int] = [1,2,3,4,5,6,7,8,9]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_primitives.nim:446]:
```suggestion
        var testVals: array[9, int] = [1,2,3,4,5,6,7,8,9]
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 22]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
   zero.setZero()
 
   var max_degree_plus_one = points.len
-
-  doAssert (max_degree_plus_one >= 2).bool() == true, "Should be interpolating for degree >= 1!"
+  doAssert max_degree_plus_one >= 2, "Should be interpolating for degree >= 1!"
 
   for k in 0 ..< points.len:
-    var point: Coord
+    var point : Coord
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:62]:
```suggestion
    var point: Coord
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 23]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
     point = points[k]
 
     var x_k : Fr
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:65]:
```suggestion
    var x_k: Fr 
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 24]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
 
   for k in 0 ..< points.len:
-    var point: Coord
+    var point : Coord
     point = points[k]
 
     var x_k : Fr 
     x_k = point.x
-    var y_k  : Fr 
+    var y_k : Fr
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:67]:
```suggestion
    var y_k: Fr 
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 25]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
+    var point : Coord
     point = points[k]
 
     var x_k : Fr 
     x_k = point.x
-    var y_k  : Fr 
+    var y_k : Fr 
     y_k = point.y
 
     var contribution : array[n,Fr]
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:70]:
```suggestion
    var contribution: array[n,Fr]
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 26]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
     var denominator : Fr
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:71]:
```suggestion
    var denominator: Fr
```

File: tests/t_ethereum_verkle_ipa_test_helper.nim [hunk 27]
```diff
@@ -48,81 +48,85 @@ func truncate* [Fr] (res: var openArray[Fr], s: openArray[Fr], to: int, n: stati
...
       var x_j : Fr 
       x_j = point.x
 
-      if j != k:
-        var differ = x_k
-        differ.diff(differ, x_j)
+      if j == k: 
+        continue
+
+      var differ : Fr
```
Comment (mratsim) [inline tests/t_ethereum_verkle_ipa_test_helper.nim:84]:
```suggestion
      var differ: Fr
```

File: constantine/eth_verkle_ipa/eth_verkle_constants.nim (hunk 13 above)
Comment (agnxsh) [reply constantine/eth_verkle_ipa/eth_verkle_constants.nim:48]: Barycentric weights store stores the weights $A_i$ from [-k,  0] and then from [k, 0], whereas inverted domain stores the k inverses. Hence, barycentric weight array is 256 * 2 long, and inverted domain is 510.

File: constantine/eth_verkle_ipa/barycentric_form.nim (hunk 9 above)
Comment (agnxsh) [reply constantine/eth_verkle_ipa/barycentric_form.nim:125]: It should be, for I'm not including 1/0 and -1/0 cases while populating the inverted domain array. Yet I can write a very specific test for this in the next pass, for ipa and multiproof tests.

**Review by mratsim — APPROVED** [review]
Thanks! will merge

**Pull request merged**